File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
private:
    int n, target;
    void backtrack(vector<int>& path, int curr, int start, vector<int>& candidates, vector<vector<int>>& ans){
        if (curr == target){
            ans.push_back(path);
            return;
        }
        
        for (int i = start; i < n; ++i){
            if (curr + candidates[i] <= target){
                path.push_back(candidates[i]);
                backtrack(path, curr + candidates[i], i, candidates, ans);
                path.pop_back();
            }
        }
    }
    
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        n = candidates.size();
        this -> target = target;
        vector<vector<int>> ans;
        vector<int> path;
        backtrack(path, 0, 0, candidates, ans);
        
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def backtrack(self, path: List[int], start: int, curr: int, candidates: List[int], target: int, ans: List[List[int]]) -> None:
        if curr == target:
            ans.append(path[:])
            return
        
        for i in range(start, len(candidates)):
            if curr + candidates[i] <= target:
                path.append(candidates[i])
                self.backtrack(path, i, curr + candidates[i], candidates, target, ans)
                path.pop()
    
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        self.backtrack([], 0, 0, candidates, target, ans)
        
        return ans